### ISSUE TITLE
Incorrect delivery of `-Wdollar-in-identifier-extension` inside stringified `asm` macros

### ISSUE THREAD

**Issue by Th3T3chn0G1t** [llvm/llvm-project#59586]
With the following program:
```c
// test.c

#define ASM_BLOCK(ops, ...) asm volatile(ops __VA_ARGS__)
#define ASM(...) #__VA_ARGS__ "\n"

int main(void) {
        ASM_BLOCK(ASM(int $3));
}
```
Compiled using `clang -Wdollar-in-identifier-extension test.c` under `clang` version `15.0.6` - the `-Wdollar-in-identifier-extension` diagnostic seems to be erroneously triggered by the `$` sign in the `ASM` macro parameter, as it is not an identifier but rather is stringified later.

My expectation would be that this situation should not trigger the diagnostic, however I could be incorrect depending on the intention of the `-Wdollar-in-identifier-extension` diagnostic.

**Comment by AaronBallman** [comment]
I looked into this one a little bit and it's likely to be tricky to silence. The issue is that we're lexing the token as an identifier when expanding `ASM_BLOCK` because the `ASM` macro expands to `__VA_ARGS__` which replays the tokens it was given. So we don't even know that we're in an `asm` block by the time we're lexing the identifier.

**Comment by Th3T3chn0G1t** [comment]
I figured it would be rather annoying considering how obscure the scenario is. Forgive my nativity, but how come identifiers are processed for this before/during the preprocess step? Would it not make more sense to issue the diagnostic after the preprocessor has completed its run?
